packages/block-library/src/table-of-contents/edit.js

```diff
@@ -59,14 +59,14 @@ export default function TableOfContentsEdit( {
 
 	// If a user clicks to a link prevent redirection and show a warning.
 	const { createWarningNotice, removeNotice } = useDispatch( noticeStore );
-	let noticeId;
+	const noticeIdRef = useRef();
 	const showRedirectionPreventedNotice = ( event ) => {
 		event.preventDefault();
 		// Remove previous warning if any, to show one at a time per block.
-		removeNotice( noticeId );
-		noticeId = `block-library/core/table-of-contents/redirection-prevented/${ instanceId }`;
+		removeNotice( noticeIdRef.current );
+		noticeIdRef.current = `block-library/core/table-of-contents/redirection-prevented/${ instanceId }`;
```

Comment:
Do we need reassignment here? The `instanceId` shouldn't change between rerenders; we can always use `clientId` if it does.

P.S. The Latest Posts block has similar logic.

https://github.com/WordPress/gutenberg/blob/acf661ce436d3e04d164dfb858853b4b796d2340/packages/block-library/src/latest-posts/edit.js#L159-L169

Reply:
Does the version in Latest Posts really work? It seems to me that the `noticeId` is reset to `undefined` on each rerender of the block's `Edit` component. When you click and the `showRedirectionPreventedNotice` handler is called, the created notice id is saved, but on the next rerender it's lost again.

Reply:
Dispatching a notice won't rerender the `Edit` component, so the `noticeId` is retained. But yes, the value might get lost.

By the way, the following snippet has the same effect and displays only a single notice.

```js
const showRedirectionPreventedNotice = ( event ) => {
	event.preventDefault();
	createWarningNotice( __( 'Links are disabled in the editor.' ), {
		id: `block-library/core/latest-posts/redirection-prevented/${ instanceId }`,
		type: 'snackbar',
	} );
};
```

Reply:
Dispatching a notice won't rerender, but if the component is rerendered for any other reason then the `noticeId` value will be lost. Adding the ref is a correct fix, and it should be applied also to the `LatestPostsEdit` component.

Reply:
I agree with @jsnajdr that the ref is the correct fix. 

I'm happy to follow up with fixing that in `LatestPostsEdit` if we all agree on it.

Reply:
Even better follow-up here: #66409